### PR TITLE
[SERVICES-311] Fix for reset password support

### PR DESCRIPTION
**Pull request by ArturKlajnerok**
This is covering changes and tests to user password hashing to meet this requirements:
- Latest reset password hash support
- Old reset password hash support
- Latest regular password hash support
- Old regular password hash support

User need to always have the hash for regular password but it's possible that hash for the reset password is not set.

/cc @Wikia/services-team

### CODE REVIEW

File: models/user.go [hunk 1]
```diff
@@ -176,25 +177,35 @@ func (user *User) MediaWikiTransformation() {
...
+}
 
-	if len(user.NewPassword) > 3 {
-		salt, passHash := ExtractHashAndSalt(user.NewPassword)
-		if passHash == HashPassword(password, salt) {
-			return true
-		}
+func (user *User) passwordMatchesHash() bool {
+	if len(user.PasswordHashed) < MinHashSize {
+		logger.GetLogger().Error("To short password hash for user: " + user.Name)
```
Comment (aquilax) [inline models/user.go]: Looks like there is a typo here `To short...` => `Too short...`

File: models/user.go [hunk 2]
```diff
@@ -176,25 +177,35 @@ func (user *User) MediaWikiTransformation() {
...
 	}
+	return user.passwordMatchesHash()
+}
 
-	if len(user.NewPassword) > 3 {
-		salt, passHash := ExtractHashAndSalt(user.NewPassword)
-		if passHash == HashPassword(password, salt) {
-			return true
-		}
+func (user *User) passwordMatchesHash() bool {
```
Comment (aquilax) [inline models/user.go]: Passing `user.Passsword` and `user.PasswordHashed` as parameters will make the funtcion easier to test. :telescope:

**Comment by aquilax** [comment]
On a side note you may consider using http://golang.org/pkg/net/mail/#ParseAddressList for validating the email address :ghost:

File: models/user.go [hunk 3]
```diff
@@ -176,25 +178,38 @@ func (user *User) MediaWikiTransformation() {
 }
 
 func (user *User) IsValidPassword(password string) bool {
-	if len(user.HashedPassword) < 3 {
-		logger.GetLogger().Error("To short hash for user: " + user.Name)
-		return false
+	user.Password = password
```
Comment (drsnyder) [inline models/user.go]: This looks odd. Why do you have to set the users password at the top of the function? I would not expect this function to have side effects or alter the state of the `user` struct.

File: models/user.go [hunk 4]
```diff
@@ -176,25 +178,38 @@ func (user *User) MediaWikiTransformation() {
...
+
+	if user.passwordMatchesHash(user.ResetPasswordHashed) {
+		return true
 	}
+	return user.passwordMatchesHash(user.PasswordHashed)
```
Comment (drsnyder) [inline models/user.go]: These 4 lines could be simplified to `return user.passwordMatchesHash(user.ResetPasswordHashed) || user.passwordMatchesHash(user.PasswordHashed)`.

Comment (ArturKlajnerok) [inline models/user.go]: Changed in: f71a009

File: models/user.go (hunk 3 above)
Comment (ArturKlajnerok) [inline models/user.go]: Solved in: 44beb1f

**Pull request merged**